File: lecode139.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <set>
using namespace std;
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        len = s.size();
        for(string str : wordDict)
            word.insert(str);
            return dfs(0,s);
    }
    bool dfs(int start,string& s){
        if (start == len) return true;
        if(vis.count(start)){
            return vis[start];
        }
        for (int end = start + 1;end < len + 1;end++){
            if(word.count(s.substr(start,end - start)) && dfs(end,s)){
                vis[start] = true;
                return true;
            }
        }
        vis[start] = false;
        return false;
    }
private:
    unordered_map<int,bool> vis;
    set<string> word;
    int len;
};
// int main(void){
//     Solution ans;
//     vector<string> str({"a","aa","aaa","aaaa","aaaaa","aaaaaa","aaaaaaa","aaaaaaaa","aaaaaaaaa","aaaaaaaaaa"});
//     printf("%d\n",ans.wordBreak(string("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab"),
//     str));
//     return 0;
// }

File: lecode139.py
from typing import List
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        word = set(wordDict)
        # 索引为start val为end
        vis = {}
        n = len(s)
        def dfs(idx : int):
            if idx == len(s):
                return True
            if  idx in vis:
                return vis[idx]
            for end in range(idx + 1,n + 1):
                if s[idx : end] in word and dfs(end):
                    vis[idx] = True
                    return True
            vis[idx] = False
            return False
        return dfs(0)

if __name__ == "__main__":
    ans = Solution()
    print(ans.wordBreak(("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab"),
    ["a","aa","aaa","aaaa","aaaaa","aaaaaa","aaaaaaa","aaaaaaaa","aaaaaaaaa","aaaaaaaaaa"]))
